libp2p/muxers/mplex/lpchannel.nim

```diff
@@ -90,20 +90,30 @@ proc reset*(s: LPChannel) {.async, gcsafe.} =
   s.readBuf = StreamSeq()
   s.pushedEof = true
 
-  let pushing = s.pushing # s.pushing changes while iterating
-  for i in 0..<pushing:
-    # Make sure to drain any ongoing pushes - there's already at least one item
-    # more in the queue already so any ongoing reads shouldn't interfere
-    # Notably, popFirst is not fair - which reader/writer gets woken up depends
-    discard await s.readQueue.popFirst()
-
-  if s.readQueue.len == 0 and s.reading:
-    # There is an active reader - we just grabbed all pushes so we need to push
-    # an EOF marker to wake it up
-    try:
-      s.readQueue.addLastNoWait(@[])
-    except CatchableError:
-      raiseAssert "We just checked the queue is empty"
+  # Essentially we need to handle the following cases
+  #
+  # - If a push was in progress but no reader is
+  # attached we need to pop the queue
+  # - If a read was in progress without without a
+  # push/data we need to push the Eof marker to
+  # notify the reader that the channel closed
+  #
+  # In all other cases, there should be a data to complete
+  # a read or enough room in the queue/buffer to complete a
+  # push.
+  #
+  # State       | Q Empty  | Q Full
+  # ------------|----------|-------
+  # Reading     | Push Eof | Na
+  # Pushing     | Na       | Pop
+  if not(s.reading and s.pushing):
```

Comment:
Maybe rather then booleans a single enum would be better at this point since it's almost a state machine.